### ISSUE TITLE
[Experiment] Layerwise Learning Rate Decay for BERTweet-large

### ISSUE THREAD

**Issue by SimoneMoreWare** [andreaGaudino/NLP_with_disaster_tweets#6]
## Summary
Layerwise learning rate decay assigns a different learning rate to each 
transformer layer — lower LR for layers closer to the input (preserve 
pre-training knowledge) and higher LR for layers closer to the output 
(adapt to the task).

## Motivation
Inspired by this notebook: https://www.kaggle.com/code/rhtsingh/utilizing-transformer-representations-efficiently
Particularly relevant for BERTweet-large (355M params) where catastrophic 
forgetting of pre-trained tweet representations is a real risk.

## Implementation
```python
def get_optimizer_grouped_parameters(model, learning_rate, 
                                      weight_decay, decay=0.9):
    layers = [model.roberta.embeddings] + \
             list(model.roberta.encoder.layer)
    layers.reverse()
    lr = learning_rate
    params = []
    for layer in layers:
        params += [{"params": layer.parameters(), "lr": lr}]
        lr *= decay
    params += [{"params": model.classifier.parameters(), 
                "lr": learning_rate}]
    return params

optimizer = torch.optim.AdamW(
    get_optimizer_grouped_parameters(model, 
                                     learning_rate=1e-5, 
                                     weight_decay=0.01,
                                     decay=0.9)
)
```

## Caveat
Requires abandoning HuggingFace Trainer and writing a manual training 
loop. Only worth pursuing if BERTweet-large results plateau.

## Expected benefit
+0.002/+0.005 F1 on Kaggle based on community results with similar models.

## References
- Notebook source: https://www.kaggle.com/code/rhtsingh/utilizing-transformer-representations-efficiently
- Paper: "Revisiting Few-sample BERT Fine-tuning" (Zhang et al. 2021)
